server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2193,13 +2210,15 @@ private void onNewEngine(Engine newEngine) {
     /**
      * called if recovery has to be restarted after network error / delay **
      */
-    public void performRecoveryRestart() throws IOException {
+    public void performRecoveryRestart(ActionListener<Void> listener) throws IOException {
         assert Thread.holdsLock(mutex) == false : "restart recovery under mutex";
-        synchronized (engineMutex) {
-            assert refreshListeners.pendingCount() == 0 : "we can't restart with pending listeners";
-            IOUtils.close(currentEngineReference.getAndSet(null));
+        engineMutex.lock();
+        assert refreshListeners.pendingCount() == 0 : "we can't restart with pending listeners";
+        Engine.close(currentEngineReference.getAndSet(null), ActionListener.runBefore(listener, () -> {
+            engineMutex.unlock();
```

Comment:
I guess this only works because the listener completes on the same thread that acquired the mutex? It's generally not safe.